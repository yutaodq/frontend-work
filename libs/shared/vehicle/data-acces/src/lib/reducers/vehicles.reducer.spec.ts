import { reducer } from '@example-app/books/reducers/books.reducer';
import * as fromVehicles from '@example-app/books/reducers/books.reducer';
import {
  VehiclesApiActions,
  VehicleActions,
  ViewVehiclePageActions,
  CollectionApiActions,
} from '@example-app/books/actions';
import { Vehicle, generateMockVehicle } from '@example-app/books/models';

describe('VehiclesReducer', () => {
  const book1 = generateMockVehicle();
  const book2 = { ...book1, id: '222' };
  const book3 = { ...book1, id: '333' };
  const initialState: fromVehicles.State = {
    ids: [book1.id, book2.id],
    entities: {
      [book1.id]: book1,
      [book2.id]: book2,
    },
    selectedVehicleId: null,
  };

  describe('undefined action', () => {
    it('should return the default state', () => {
      const result = reducer(undefined, {} as any);

      expect(result).toMatchSnapshot();
    });
  });

  describe('SEARCH_COMPLETE & LOAD_SUCCESS', () => {
    type VehiclesActions =
      | typeof VehiclesApiActions.searchSuccess
      | typeof CollectionApiActions.loadVehiclesSuccess;
    function noExistingVehicles(
      action: VehiclesActions,
      booksInitialState: any,
      books: Vehicle[]
    ) {
      const createAction = action({ books });

      const result = reducer(booksInitialState, createAction);

      expect(result).toMatchSnapshot();
    }

    function existingVehicles(
      action: VehiclesActions,
      booksInitialState: any,
      books: Vehicle[]
    ) {
      // should not replace existing books
      const differentVehicle2 = { ...books[0], foo: 'bar' };
      const createAction = action({ books: [books[1], differentVehicle2] });

      const expectedResult = {
        ids: [...booksInitialState.ids, books[1].id],
        entities: {
          ...booksInitialState.entities,
          [books[1].id]: books[1],
        },
        selectedVehicleId: null,
      };

      const result = reducer(booksInitialState, createAction);

      expect(result).toMatchSnapshot();
    }

    it('should add all books in the payload when none exist', () => {
      noExistingVehicles(VehiclesApiActions.searchSuccess, initialState, [
        book1,
        book2,
      ]);

      noExistingVehicles(CollectionApiActions.loadVehiclesSuccess, initialState, [
        book1,
        book2,
      ]);
    });

    it('should add only books when books already exist', () => {
      existingVehicles(VehiclesApiActions.searchSuccess, initialState, [
        book2,
        book3,
      ]);

      existingVehicles(CollectionApiActions.loadVehiclesSuccess, initialState, [
        book2,
        book3,
      ]);
    });
  });

  describe('LOAD', () => {
    const expectedResult = {
      ids: [book1.id],
      entities: {
        [book1.id]: book1,
      },
      selectedVehicleId: null,
    };

    it('should add a single vehicle, if the vehicle does not exist', () => {
      const action = VehicleActions.loadVehicle({ vehicle: book1 });

      const result = reducer(fromVehicles.initialState, action);

      expect(result).toMatchSnapshot();
    });

    it('should return the existing state if the vehicle exists', () => {
      const action = VehicleActions.loadVehicle({ vehicle: book1 });

      const result = reducer(expectedResult, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('SELECT', () => {
    it('should set the selected vehicle id on the state', () => {
      const action = ViewVehiclePageActions.selectVehicle({ id: book1.id });

      const result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('Selectors', () => {
    describe('selectId', () => {
      it('should return the selected id', () => {
        const result = fromVehicles.selectId({
          ...initialState,
          selectedVehicleId: book1.id,
        });

        expect(result).toMatchSnapshot();
      });
    });
  });
});
