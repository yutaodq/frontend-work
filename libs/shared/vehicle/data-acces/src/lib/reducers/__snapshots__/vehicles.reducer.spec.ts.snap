// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VehiclesReducer LOAD should add a single vehicle, if the vehicle does not exist 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
  },
  "ids": Array [
    "1",
  ],
  "selectedVehicleId": null,
}
`;

exports[`VehiclesReducer LOAD should return the existing state if the vehicle exists 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
  },
  "ids": Array [
    "1",
  ],
  "selectedVehicleId": null,
}
`;

exports[`VehiclesReducer SEARCH_COMPLETE & LOAD_SUCCESS should add all books in the payload when none exist 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
    "222": Object {
      "id": "222",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
  },
  "ids": Array [
    "1",
    "222",
  ],
  "selectedVehicleId": null,
}
`;

exports[`VehiclesReducer SEARCH_COMPLETE & LOAD_SUCCESS should add all books in the payload when none exist 2`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
    "222": Object {
      "id": "222",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
  },
  "ids": Array [
    "1",
    "222",
  ],
  "selectedVehicleId": null,
}
`;

exports[`VehiclesReducer SEARCH_COMPLETE & LOAD_SUCCESS should add only books when books already exist 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
    "222": Object {
      "id": "222",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
    "333": Object {
      "id": "333",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
  },
  "ids": Array [
    "1",
    "222",
    "333",
  ],
  "selectedVehicleId": null,
}
`;

exports[`VehiclesReducer SEARCH_COMPLETE & LOAD_SUCCESS should add only books when books already exist 2`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
    "222": Object {
      "id": "222",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
    "333": Object {
      "id": "333",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
  },
  "ids": Array [
    "1",
    "222",
    "333",
  ],
  "selectedVehicleId": null,
}
`;

exports[`VehiclesReducer SELECT should set the selected vehicle id on the state 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": "1",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
    "222": Object {
      "id": "222",
      "volumeInfo": Object {
        "authors": Array [
          "author",
        ],
        "averageRating": 3,
        "description": "description",
        "imageLinks": Object {
          "smallThumbnail": "string",
          "thumbnail": "string",
        },
        "publishDate": "",
        "publisher": "publisher",
        "ratingsCount": 5,
        "subtitle": "subtitle",
        "title": "title",
      },
    },
  },
  "ids": Array [
    "1",
    "222",
  ],
  "selectedVehicleId": "1",
}
`;

exports[`VehiclesReducer Selectors selectId should return the selected id 1`] = `"1"`;

exports[`VehiclesReducer undefined action should return the default state 1`] = `
Object {
  "entities": Object {},
  "ids": Array [],
  "selectedVehicleId": null,
}
`;
